File: packages/MeeshySDK/Sources/MeeshyUI/Auth/SignupForm.swift
import Foundation
import MeeshySDK

/// Ce que l'écran d'inscription SAIT, sans rien de ce qu'il MONTRE.
///
/// Une `struct` pure : quatre saisies (nom, e-mail, téléphone, mot de passe),
/// le pays de l'indicatif, les deux rangs du Prisme, leurs règles locales, et la
/// composition de la charge `POST /auth/register`. Aucun réseau, aucun
/// singleton, aucune horloge — donc entièrement éprouvable sans simulateur.
///
/// ### Pourquoi au SDK, et non côté app
///
/// Le test du grain de `packages/MeeshySDK/CLAUDE.md` la place ici sur ses
/// quatre questions : elle prend des paramètres OPAQUES (une `Locale`), ne lit
/// aucun singleton nommé Meeshy, ne résout aucune URL et n'applique aucune
/// cascade de replis UX. Ce qu'elle encode n'est pas une décision produit mais
/// le CONTRAT de `RegisterRequest`, un type du SDK : quels champs partent, et
/// lesquels sont absents quand ils sont vides. La décision produit — quand
/// envoyer, quoi afficher d'un refus, où mener après — vit dans
/// `SignupViewModel`, côté app.
///
/// La ligne se lit ainsi : **la forme SAIT composer la charge ; elle ne sait pas
/// qu'un bouton existe.**
public struct SignupForm: Equatable {

    // MARK: - Saisies

    /// Le pseudo et le nom affiché TAPÉS (#6479, refaits par #7897).
    ///
    /// `nil` = jamais touché : le champ SUIT l'adresse en direct. Une chaîne =
    /// la saisie, qui gagne ; vidée, la dérivation reparaît en filigrane et
    /// repart. L'écran montre `effectiveUsername` / `effectiveDisplayName` et
    /// écrit ici. Miroir de `SignupFormState` (`apps/web/src/lib/signup-form.ts`).
    public var username: String?
    public var displayName: String?
    public var email: String
    /// Les chiffres SEULS, sans indicatif : l'indicatif vient de `country`.
    public var phoneDigits: String
    public var password: String
    /// Le pays de l'indicatif téléphonique, pré-sélectionné depuis la locale.
    public var country: CountryCode
    /// Rang 1 du Prisme — la langue dans laquelle l'utilisateur LIRA Meeshy.
    public var systemLanguage: String
    /// Rang 2, déduit de la RÉGION et jamais montré : un francophone au Canada
    /// lit en français, et l'anglais reste servi quand le français manque.
    public var regionalLanguage: String

    // MARK: - Règles locales
    //
    // Miroirs de `AuthSchemas.register` (packages/shared/utils/validation.ts).
    // Elles ne dupliquent PAS le serveur : elles évitent d'envoyer une charge
    // dont on sait déjà qu'elle sera refusée. Le serveur reste l'arbitre.

    /// Longueur maximale d'un nom affiché — **100**, la valeur du schéma serveur
    /// (`AuthSchemas.register.displayName`, `z.string().min(1).max(100)`).
    ///
    /// Ce n'est PAS la limite de `firstName`/`lastName` (50 chacun) : ceux-là ne
    /// partent plus, la passerelle les dérive. Recopier 50 ici rendrait le
    /// client plus STRICT que le serveur — un refus local pour une saisie que la
    /// passerelle aurait acceptée, ce qui est le défaut le plus difficile à
    /// diagnostiquer d'une validation miroir : rien ne rougit nulle part.
    public static let displayNameMaxLength = 100

    /// Longueur minimale du mot de passe — **6**, miroir de
    /// `PASSWORD_MIN_LENGTH` (`packages/shared/utils/validation.ts`).
    ///
    /// C'est la SEULE source du minimum côté client : le placeholder du champ
    /// l'annonce, la validation locale l'applique. Le dépôt n'emploie nulle part
    /// `UITextInputPasswordRules`, et ce lot ne l'introduit pas — la règle n'y
    /// serait pas plus vraie, seulement écrite une deuxième fois, dans une API
    /// que la cible iOS 16 de l'app ne sert pas partout.
    public static let passwordMinLength = 6

    /// Miroir du pattern serveur `personNamePatternSource`
    /// (`^(?=.*\p{L})[\p{L}\p{M}\s'’ʼ.-]+$`) : lettres (accents et marques
    /// combinantes incluses — `CharacterSet.letters` couvre L* et M*), espaces,
    /// apostrophes droite ET typographiques (le clavier iOS insère `’`), points
    /// et tirets ; au moins une lettre ; `displayNameMaxLength` caractères au plus.
    ///
    /// Un nom FOURNI tient le pattern et la borne ; un champ VIDE est valide,
    /// parce qu'il est LÉGITIME (#6441, suite de #6424) — la passerelle le
    /// DÉRIVE alors de la partie locale de l'adresse (`displayNameDepuisEmail`,
    /// `services/gateway/src/services/auth/registration-identity.ts`). C'est le
    /// même arbitrage que `isPasswordValid` ci-dessous, et le doc-comment de
    /// `displayNameMaxLength` nomme déjà le défaut qu'il écarte : un refus
    /// local pour une charge que la passerelle ACCEPTE, et rien ne rougit
    /// nulle part.
    public static func isDisplayNameValid(_ value: String) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return true }
        guard trimmed.count <= displayNameMaxLength else { return false }
        guard trimmed.unicodeScalars.contains(where: { CharacterSet.letters.contains($0) }) else { return false }
        let allowed = CharacterSet.letters
            .union(.whitespaces)
            .union(CharacterSet(charactersIn: "'’ʼ.-"))
        return trimmed.unicodeScalars.allSatisfy { allowed.contains($0) }
    }

    /// Rapprochement du `z.email` serveur : « a@b » et « alice@com. » le
    /// passaient quand la règle était « contient @ et . ».
    public static func isEmailValid(_ value: String) -> Bool {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.range(of: #"^[^@\s]+@[^@\s]+\.[A-Za-z]{2,}$"#, options: .regularExpression) != nil
    }

    /// Un mot de passe FOURNI doit atteindre la borne ; un champ VIDE est
    /// valide, parce qu'il est LÉGITIME (#6424).
    ///
    /// Le compte naît alors sans mot de passe, et sa porte est le lien magique.
    /// Garder l'ancienne règle ici rendrait le client plus strict que le
    /// serveur — le défaut que le doc-comment de `displayNameMaxLength` nomme
    /// juste au-dessus : un refus local pour une charge que la passerelle
    /// accepte, et rien ne rougit nulle part.
    public static func isPasswordValid(_ value: String) -> Bool {
        value.isEmpty || value.count >= passwordMinLength
    }

    /// `true` quand l'utilisateur a réellement TAPÉ un mot de passe.
    ///
    /// Distinct de `isPasswordValid`, qui répond « cette saisie est-elle
    /// acceptable ». Celle-ci répond « y a-t-il une saisie », et c'est elle qui
    /// décide si la clé part dans la charge.
    public static func hasPassword(_ value: String) -> Bool { !value.isEmpty }

    /// `true` quand l'utilisateur a réellement TAPÉ un nom affiché.
    ///
    /// Jumelle de `hasPassword` : `isDisplayNameValid` répond « cette saisie
    /// est-elle acceptable », celle-ci « y a-t-il une saisie » — et c'est elle
    /// qui décide si la clé part dans la charge.
    public static func hasDisplayName(_ value: String) -> Bool {
        !value.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    public var isDisplayNameValid: Bool { Self.isDisplayNameValid(displayName ?? "") }
    public var isEmailValid: Bool { Self.isEmailValid(email) }
    public var isPasswordValid: Bool { Self.isPasswordValid(password) }
    public var hasPassword: Bool { Self.hasPassword(password) }
    public var hasDisplayName: Bool { Self.hasDisplayName(displayName ?? "") }

    /// LE PSEUDO QUI PARTIRA — tapé s'il l'a été, tiré de l'ADRESSE sinon
    /// (#6479, #7897 : « le pseudo rempli à partir de l'email »).
    ///
    /// `resoudreUsername` (`registration.service.ts`) emploie tel quel un
    /// pseudo fourni. Rend `""` quand la dérivation retombe sur le RECOURS
    /// (`user`) : l'envoyer garantirait une collision.
    public var effectiveUsername: String {
        let tape = (username ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        if !tape.isEmpty { return tape }
        let derive = RegistrationIdentity.pseudoRacine(displayName: nil, email: email)
        return derive == RegistrationIdentity.pseudoDeSecours ? "" : derive
    }

    /// Le nom affiché qui partira — tapé s'il l'a été, dérivé de l'adresse sinon.
    public var effectiveDisplayName: String {
        let tape = (displayName ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        return tape.isEmpty ? RegistrationIdentity.displayNameDepuisEmail(email) : tape
    }

    /// L'IDENTITÉ EST DÉFINIE (#7897) — nom affiché ET pseudo existent et
    /// tiennent leurs bornes. Directive porteur : « c'est quand tout est
    /// défini qu'on active le champ mot de passe ».
    public var isIdentityDefined: Bool {
        isEmailValid
            && !effectiveDisplayName.isEmpty
            && Self.isDisplayNameValid(effectiveDisplayName)
            && !effectiveUsername.isEmpty
    }

    /// Un numéro FOURNI doit être plausible (#6479) ; un champ VIDE reste
    /// valide — il n'est pas requis. Troisième champ à porter cette forme,
    /// après le mot de passe et le nom affiché.
    public var isPhoneValid: Bool { PhonePlausibility.isPlausible(phoneDigits) }

    /// Le MOTIF du refus, pour que l'écran dise quoi corriger — « numéro
    /// invalide » n'apprend rien à qui a tapé le sien de travers.
    public var phoneRefusal: PhonePlausibility.Refusal? { PhonePlausibility.refusal(phoneDigits) }

    /// Le bouton s'active dès que l'ADRESSE est valide — et que le nom affiché,
    /// le mot de passe et le NUMÉRO, s'ils ont été tapés, tiennent leurs bornes
    /// (#6441, #6479).
    ///
    /// L'adresse est le SEUL champ requis, exactement comme le
    /// `required: ['email']` du schéma partagé : ni le nom, ni le téléphone, ni
    /// le mot de passe ne sont exigés par la passerelle. Rien ici ne dépend du
    /// réseau — aucun appel de disponibilité ne précède l'envoi.
    public var canSubmit: Bool {
        isDisplayNameValid && isEmailValid && isPasswordValid && isPhoneValid
    }

    // MARK: - Téléphone

    /// Les chiffres saisis, débarrassés de tout ce qui n'en est pas (espaces,
    /// points, parenthèses collées depuis un carnet d'adresses).
    public var normalizedPhoneDigits: String {
        phoneDigits.filter(\.isNumber)
    }

    public var hasPhone: Bool { !normalizedPhoneDigits.isEmpty }

    // MARK: - Composition de la charge

    /// La charge exacte de `POST /auth/register`.
    ///
    /// Sept clés au plus, et jamais `username` / `firstName` / `lastName` : la
    /// passerelle les dérive de `displayName`. Le couple téléphone est TOUT ou
    /// RIEN — un `phoneCountryCode` sans numéro décrirait un pays qui ne
    /// qualifie rien.
    ///
    /// Les chiffres partent TELS QUE TAPÉS, avec leur pays — jamais préfixés
    /// de l'indicatif. Un « 06 12 34 56 78 » français composé en `+33` +
    /// `0612345678` porte un préfixe national que l'E.164 ne connaît pas, et
    /// retirer ce zéro soi-même se trompe dès l'Italie, dont les fixes le
    /// GARDENT. La passerelle normalise avec `phoneCountryCode`
    /// (libphonenumber, `normalizePhoneWithCountry`) : c'est son site unique,
    /// et c'est exactement ce que le web v3 lui remet — le champ tel quel et
    /// l'ISO du sélecteur.
    public func registerRequest() -> RegisterRequest {
        let phone = hasPhone ? normalizedPhoneDigits : nil
        return RegisterRequest(
            // CE QUE L'ÉCRAN MONTRE EST CE QU'IL ENVOIE (#6479) — dérivé ou
            // tapé, peu importe : dès qu'une valeur existe, la passerelle n'a
            // plus à en inventer une. Les deux clés restent `nil` quand il n'y
            // a réellement rien, et `nil` — jamais `""`, que `minLength: 1`
            // refuserait.
            displayName: effectiveDisplayName.isEmpty ? nil : effectiveDisplayName,
            email: email.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
            // `nil`, jamais `""` (#6424) : la passerelle lit l'ABSENCE de la
            // clé pour créer un compte sans mot de passe. Une chaîne vide
            // serait une VALEUR, refusée par la borne de longueur — le
            // formulaire échouerait précisément dans le cas qu'il vient
            // d'ouvrir.
            password: hasPassword ? password : nil,
            phoneNumber: phone,
            phoneCountryCode: phone == nil ? nil : country.id,
            systemLanguage: systemLanguage,
            regionalLanguage: regionalLanguage,
            username: effectiveUsername.isEmpty ? nil : effectiveUsername
        )
    }

    // MARK: - Défauts déduits d'une locale

    /// Région → langue régionale (rang 2 du Prisme). Repris tel quel du wizard
    /// remplacé : la table est la seule chose qu'il avait de juste sur ce point.
    static let regionLanguageMap: [String: String] = [
        "CM": "fr", "FR": "fr", "BE": "fr", "CH": "fr", "CA": "fr", "SN": "fr", "CI": "fr", "CD": "fr", "MG": "fr",
        "US": "en", "GB": "en", "AU": "en", "NZ": "en", "IE": "en", "ZA": "en", "NG": "en", "GH": "en", "KE": "en",
        "ES": "es", "MX": "es", "AR": "es", "CO": "es", "CL": "es", "PE": "es",
        "DE": "de", "AT": "de",
        "IT": "it",
        "PT": "pt", "BR": "pt",
        "SA": "ar", "AE": "ar", "EG": "ar", "MA": "ar", "DZ": "ar", "TN": "ar",
        "CN": "zh", "TW": "zh", "HK": "zh",
        "JP": "ja",
        "KR": "ko",
        "RU": "ru",
        "TR": "tr",
        "NL": "nl",
        "PL": "pl",
        "SE": "sv",
        "IN": "hi",
        "TH": "th",
        "VN": "vi",
        "UA": "uk",
        "RO": "ro",
    ]

    /// Repli ultime quand la locale ne dit rien d'exploitable.
    public static let fallbackLanguage = "fr"

    /// La langue de l'appareil si Meeshy la sert, sinon `fr`.
    public static func defaultSystemLanguage(for locale: Locale) -> String {
        guard let code = locale.language.languageCode?.identifier.lowercased(),
              LanguageData.supportedCodeSet.contains(code) else { return fallbackLanguage }
        return code
    }

    /// La langue de la RÉGION, quand elle diffère du rang 1 ; sinon l'anglais —
    /// ou le français si le rang 1 est déjà l'anglais.
    ///
    /// Elle ne se montre pas : deux rangs suffisent au Prisme, et faire choisir
    /// le second coûterait un écran pour un réglage que personne ne demande.
    public static func defaultRegionalLanguage(for locale: Locale, systemLanguage: String) -> String {
        if let region = locale.region?.identifier.uppercased(),
           let regional = regionLanguageMap[region],
           LanguageData.supportedCodeSet.contains(regional),
           regional != systemLanguage {
            return regional
        }
        return systemLanguage != "en" ? "en" : fallbackLanguage
    }

    /// Le pays de l'appareil, à défaut le premier de la liste (la France, tête
    /// de `CountryPicker.countries` par ordre de priorité).
    public static func defaultCountry(for locale: Locale) -> CountryCode {
        let fallback = CountryPicker.countries[0]
        guard let region = locale.region?.identifier.uppercased() else { return fallback }
        return CountryPicker.countries.first { $0.id == region } ?? fallback
    }

    /// Une forme vierge dont les défauts SONT déjà justes : pays, langue lue,
    /// langue régionale. Rien à configurer avant de commencer à taper.
    public init(locale: Locale = .current) {
        let system = Self.defaultSystemLanguage(for: locale)
        self.username = nil
        self.displayName = nil
        self.email = ""
        self.phoneDigits = ""
        self.password = ""
        self.country = Self.defaultCountry(for: locale)
        self.systemLanguage = system
        self.regionalLanguage = Self.defaultRegionalLanguage(for: locale, systemLanguage: system)
    }

    /// Initialiseur complet — les suites l'emploient pour poser un état sans
    /// rejouer la saisie champ par champ.
    public init(
        username: String? = nil,
        displayName: String?,
        email: String,
        phoneDigits: String = "",
        password: String,
        country: CountryCode,
        systemLanguage: String,
        regionalLanguage: String
    ) {
        self.username = username
        self.displayName = displayName
        self.email = email
        self.phoneDigits = phoneDigits
        self.password = password
        self.country = country
        self.systemLanguage = systemLanguage
        self.regionalLanguage = regionalLanguage
    }

    // MARK: - Libellé de la langue lue

    /// « Français », « English »… — le nom NATIF de la langue de lecture, tel
    /// que la pastille l'affiche. Le nom natif, et non traduit : c'est celui que
    /// son locuteur reconnaît, quelle que soit la langue du reste de l'écran.
    public var systemLanguageNativeName: String {
        LanguageData.info(for: systemLanguage)?.nativeName ?? systemLanguage.uppercased()
    }

    public var systemLanguageFlag: String {
        LanguageData.info(for: systemLanguage)?.flag ?? "🌐"
    }
}

File: apps/web/src/lib/signup-form.ts
import { personNamePatternSource, registerRequestSchema } from '@meeshy/shared/types/api-schemas/auth';
import { isSupportedLanguage } from '@meeshy/shared/utils/languages';
import { phoneImplausibility, type PhoneImplausibility } from '@meeshy/shared/utils/phone-plausibility';
import {
  PSEUDO_DE_SECOURS,
  displayNameDepuisEmail,
  pseudoRacine,
} from '@meeshy/shared/utils/registration-identity';

import type { RegisterBody } from './api/auth';
import { countryOf, type Country } from './countries';

/**
 * LE MODÈLE D'INSCRIPTION (#5555, E4) — miroir pur de `SignupForm.swift`
 * (`packages/MeeshySDK/Sources/MeeshyUI/Auth/SignupForm.swift`), avec les
 * SOURCES PARTAGÉES à la place des constantes recopiées : `PASSWORD_MIN` est
 * LU sur `registerRequestSchema`, jamais un littéral local — la borne serveur
 * a déjà divergé une fois (#3629, 6 → 12) sans que le miroir Swift le
 * remarque (doc-comment `SignupForm.swift:59-67`, toujours à 6).
 *
 * Aucun réseau, aucun singleton, aucune horloge : entièrement pur, comme son
 * modèle iOS. La décision produit (quand envoyer, où poser un refus) vit dans
 * `view/auth-feedback.ts` et l'écran, jamais ici.
 */

export type SignupFormState = {
  /**
   * Le pseudo et le nom affiché TAPÉS (#6479, refaits par #7897).
   *
   * `null` = jamais touché : le champ SUIT l'adresse en direct. Une chaîne =
   * la saisie, qui gagne ; vidée, la dérivation reparaît en filigrane et
   * repart. L'écran montre `effectiveUsername` / `effectiveDisplayName` et
   * écrit ici : sans cette séparation, taper une lettre dans l'adresse
   * écraserait un pseudo choisi à la main.
   */
  readonly username: string | null;
  readonly displayName: string | null;
  readonly email: string;
  /** Les chiffres SEULS, sans indicatif — l'indicatif vient de `country`. */
  readonly phoneDigits: string;
  readonly password: string;
  readonly country: Country;
  /** Rang 1 du Prisme — la langue dans laquelle l'utilisateur LIRA Meeshy. */
  readonly systemLanguage: string;
  /** Rang 2, déduit de la RÉGION et jamais montré. */
  readonly regionalLanguage: string;
};

/** Longueur maximale d'un nom affiché — miroir de `displayNameProperty`
 * (`registerRequestSchema`, `types/api-schemas/auth.ts`). */
export const DISPLAY_NAME_MAX = registerRequestSchema.properties.displayName.maxLength;

/** Longueur minimale du mot de passe — LA source, jamais un littéral local
 * (#3629 : la borne serveur est passée de 6 à 12 sans que le miroir iOS le
 * suive ; ce module ne peut pas rejouer ce défaut, il LIT le schéma). */
export const PASSWORD_MIN = registerRequestSchema.properties.password.minLength;

/** Compilé depuis la source PARTAGÉE — la même regex que le schéma Ajv du
 * serveur et le Zod de `AuthSchemas.register` compilent tous deux. */
const PERSON_NAME_PATTERN = new RegExp(personNamePatternSource, 'u');

/**
 * Un nom affiché FOURNI doit tenir le pattern et la borne ; un champ VIDE est
 * valide, parce qu'il est LÉGITIME (#6441, suite de #6424) — la passerelle le
 * DÉRIVE alors de la partie locale de l'adresse (`displayNameDepuisEmail`,
 * `services/gateway/src/services/auth/registration-identity.ts`).
 *
 * C'est MOT POUR MOT l'arbitrage rendu pour le mot de passe trente lignes plus
 * bas, et le doc-comment de `DISPLAY_NAME_MAX` nomme déjà le défaut qu'il
 * écarte : garder l'ancienne règle rendrait le client plus STRICT que le
 * serveur — un refus local pour une charge que la passerelle ACCEPTE, et rien
 * ne rougit nulle part. La moitié « mot de passe » de #6424 a été livrée sans
 * celle-ci : l'écran annonçait « adresse seule » et exigeait toujours un nom.
 */
export function isDisplayNameValid(value: string | null): boolean {
  const trimmed = (value ?? '').trim();
  if (trimmed.length === 0) return true;
  if (trimmed.length > DISPLAY_NAME_MAX) return false;
  return PERSON_NAME_PATTERN.test(trimmed);
}

/** `true` quand un nom affiché a réellement été TAPÉ. Distinct de
 * `isDisplayNameValid`, qui répond « cette saisie est-elle acceptable » :
 * c'est celle-ci qui décide si la clé part dans la charge. */
export function hasDisplayName(value: string | null): boolean {
  return (value ?? '').trim().length > 0;
}

/** Rapprochement du `z.email`/`format: 'email'` serveur : « a@b » (pas de TLD)
 * est refusé, comme il l'est côté serveur. */
export function isEmailValid(value: string): boolean {
  const trimmed = value.trim();
  return /^[^@\s]+@[^@\s]+\.[A-Za-z]{2,}$/.test(trimmed);
}

/**
 * Un mot de passe FOURNI doit atteindre la borne ; un champ VIDE est valide,
 * parce qu'il est LÉGITIME (#6424) — le compte naît alors sans mot de passe et
 * sa porte est le lien magique.
 *
 * Garder l'ancienne règle rendrait le client plus STRICT que le serveur : un
 * refus local pour une charge que la passerelle accepte, et rien ne rougit
 * nulle part (§ `DISPLAY_NAME_MAX`, même piège).
 */
export function isPasswordValid(value: string): boolean {
  return value.length === 0 || value.length >= PASSWORD_MIN;
}

/** `true` quand un mot de passe a réellement été TAPÉ. Distinct de
 * `isPasswordValid`, qui répond « cette saisie est-elle acceptable » : c'est
 * celle-ci qui décide si la clé part dans la charge. */
export function hasPassword(value: string): boolean {
  return value.length > 0;
}

/** Le bouton s'active dès que l'ADRESSE est valide — et que le nom affiché, le
 * mot de passe et le NUMÉRO, s'ils ont été tapés, tiennent leurs bornes
 * (#6441, #6479).
 * L'adresse est le SEUL champ requis, comme `required: ['email']` du schéma
 * partagé : ni le nom, ni le téléphone, ni le mot de passe ne sont exigés par
 * la passerelle (§ SignupForm.swift, le miroir qui porte la même loi). */
export function canSubmit(form: SignupFormState): boolean {
  return (
    isDisplayNameValid(form.displayName) &&
    isEmailValid(form.email) &&
    isPasswordValid(form.password) &&
    isPhoneValid(form.phoneDigits)
  );
}

/**
 * Un numéro FOURNI doit être plausible (#6479) ; un champ VIDE reste valide —
 * le numéro n'est pas requis (#6424). Troisième champ à porter cette forme,
 * après le mot de passe et le nom affiché.
 *
 * La loi vit dans `@meeshy/shared/utils/phone-plausibility`, jamais ici : la
 * passerelle devra la partager pour que le refus soit le MÊME des deux côtés,
 * et une jumelle locale rendrait cette convergence impossible.
 */
export function isPhoneValid(phoneDigits: string): boolean {
  return phoneImplausibility(phoneDigits) === null;
}

/** Le MOTIF du refus, pour que l'écran dise quoi corriger — « numéro
 * invalide » n'apprend rien à qui a tapé le sien de travers. */
export function phoneRefusal(phoneDigits: string): PhoneImplausibility | null {
  return phoneImplausibility(phoneDigits);
}

/**
 * LE PSEUDO QUI PARTIRA — tapé s'il l'a été, tiré de l'ADRESSE sinon (#6479,
 * #7897 : « le pseudo rempli à partir de l'email »).
 *
 * Directive porteur : « à partir du moment où un champ username est rempli, la
 * passerelle n'a plus rien à créer ». `resoudreUsername`
 * (`registration.service.ts`) emploie tel quel un pseudo fourni.
 *
 * Rend `''` quand la dérivation retombe sur le RECOURS (`user`) : envoyer
 * `user` garantirait une collision — la passerelle cherche alors elle-même.
 */
export function effectiveUsername(form: SignupFormState): string {
  const tape = (form.username ?? '').trim();
  if (tape !== '') return tape;
  const derive = pseudoRacine({ email: form.email });
  return derive === PSEUDO_DE_SECOURS ? '' : derive;
}

/** Le nom affiché qui partira — tapé s'il l'a été, tiré de l'adresse sinon. */
export function effectiveDisplayName(form: SignupFormState): string {
  const tape = (form.displayName ?? '').trim();
  return tape !== '' ? tape : displayNameDepuisEmail(form.email);
}

/**
 * L'IDENTITÉ EST DÉFINIE (#7897) — nom affiché ET pseudo existent et tiennent
 * leurs bornes. Directive porteur : « c'est quand tout est défini qu'on active
 * le champ mot de passe ».
 */
export function isIdentityDefined(form: SignupFormState): boolean {
  return (
    isEmailValid(form.email) &&
    effectiveDisplayName(form) !== '' &&
    isDisplayNameValid(effectiveDisplayName(form)) &&
    effectiveUsername(form) !== ''
  );
}

/** Les chiffres saisis, débarrassés de tout ce qui n'en est pas. */
export function normalizedPhoneDigits(phoneDigits: string): string {
  return phoneDigits.replace(/\D/g, '');
}

/**
 * UN NUMÉRO A-T-IL ÉTÉ DONNÉ ? — la MÊME question que la charge se pose
 * (`composeRegisterBody`) et que l'alerte d'une inscription sans numéro pose
 * à l'écran (#8040) : l'alerte ne peut pas paraître pour une saisie que la
 * charge enverrait, ni se taire pour une saisie qu'elle omettrait.
 */
export function hasPhoneNumber(form: SignupFormState): boolean {
  return normalizedPhoneDigits(form.phoneDigits).length > 0;
}

/**
 * La charge EXACTE de `POST /auth/register` (`register.ts:133`) — sept clés
 * au plus, jamais `username` / `firstName` / `lastName` (la passerelle les
 * dérive de `displayName`, #5218). Le couple téléphone est TOUT ou RIEN : un
 * numéro sans pays ne qualifierait rien.
 */
export function composeRegisterBody(form: SignupFormState): RegisterBody {
  const digits = normalizedPhoneDigits(form.phoneDigits);
  return {
    // OMISE quand le champ est vide, jamais `''` : `displayNameProperty` porte
    // `minLength: 1` — une chaîne vide serait une VALEUR, refusée par la borne,
    // et l'inscription échouerait dans le cas même qu'elle ouvre (#6441).
    // Ce que l'écran MONTRE est ce qu'il ENVOIE (#6479) — dérivé ou tapé, peu
    // importe : dès qu'une valeur existe, la passerelle n'a plus à en inventer
    // une. Les deux clés restent OMISES quand il n'y a réellement rien.
    ...(effectiveUsername(form) !== '' ? { username: effectiveUsername(form) } : {}),
    ...(effectiveDisplayName(form) !== '' ? { displayName: effectiveDisplayName(form) } : {}),
    email: form.email.trim().toLowerCase(),
    // `undefined` par OMISSION, jamais `''` (#6424) — même raison que le couple
    // téléphone une ligne plus bas : une clé présente à valeur vide décrit
    // quelque chose qui n'a pas été demandé.
    ...(hasPassword(form.password) ? { password: form.password } : {}),
    ...(hasPhoneNumber(form) ? { phoneNumber: digits, phoneCountryCode: form.country.id } : {}),
    systemLanguage: form.systemLanguage,
    regionalLanguage: form.regionalLanguage,
  };
}

// --- Défauts déduits d'une locale (miroir SignupForm.swift:148-206) --------

/** Miroir de `SignupForm.regionLanguageMap` — reprise telle quelle, la seule
 * chose que le wizard remplacé avait de juste sur ce point. */
const REGION_LANGUAGE: Readonly<Record<string, string>> = {
  CM: 'fr', FR: 'fr', BE: 'fr', CH: 'fr', CA: 'fr', SN: 'fr', CI: 'fr', CD: 'fr', MG: 'fr',
  US: 'en', GB: 'en', AU: 'en', NZ: 'en', IE: 'en', ZA: 'en', NG: 'en', GH: 'en', KE: 'en',
  ES: 'es', MX: 'es', AR: 'es', CO: 'es', CL: 'es', PE: 'es',
  DE: 'de', AT: 'de',
  IT: 'it',
  PT: 'pt', BR: 'pt',
  SA: 'ar', AE: 'ar', EG: 'ar', MA: 'ar', DZ: 'ar', TN: 'ar',
  CN: 'zh', TW: 'zh', HK: 'zh',
  JP: 'ja',
  KR: 'ko',
  RU: 'ru',
  TR: 'tr',
  NL: 'nl',
  PL: 'pl',
  SE: 'sv',
  IN: 'hi',
  TH: 'th',
  VN: 'vi',
  UA: 'uk',
  RO: 'ro',
};

const FALLBACK_LANGUAGE = 'fr';

export type DefaultLanguages = { readonly systemLanguage: string; readonly regionalLanguage: string };

/**
 * Défauts déduits d'une `Intl.Locale` (ou de sa forme texte, ex. `"fr-FR"`) —
 * miroir de `SignupForm.defaultSystemLanguage` / `defaultRegionalLanguage`.
 * Le rang 2 ne se montre jamais : c'est le témoin de RANG de ce module (le
 * témoin s'écrit sur `en-US` — rang 2, jamais sur le rang 1).
 */
export function defaultLanguages(locale: string): DefaultLanguages {
  const [languageTag, regionTag] = locale.split('-');
  const language = (languageTag ?? '').toLowerCase();
  const systemLanguage = isSupportedLanguage(language) ? language : FALLBACK_LANGUAGE;

  const region = (regionTag ?? '').toUpperCase();
  const regional = REGION_LANGUAGE[region];
  const regionalLanguage =
    regional && isSupportedLanguage(regional) && regional !== systemLanguage
      ? regional
      : systemLanguage !== 'en'
        ? 'en'
        : FALLBACK_LANGUAGE;

  return { systemLanguage, regionalLanguage };
}

/** Le pays de l'appareil, à défaut le premier de la liste (la France, tête de
 * `COUNTRIES` par ordre de priorité — miroir `SignupForm.defaultCountry`). */
export function defaultCountry(locale: string): Country {
  const [, regionTag] = locale.split('-');
  const region = (regionTag ?? '').toUpperCase();
  return countryOf(region) ?? countryOf('FR')!;
}

/** Une forme vierge dont les défauts SONT déjà justes — miroir
 * `SignupForm.init(locale:)`. */
export function emptySignupForm(locale: string): SignupFormState {
  const { systemLanguage, regionalLanguage } = defaultLanguages(locale);
  return {
    username: null,
    displayName: null,
    email: '',
    phoneDigits: '',
    password: '',
    country: defaultCountry(locale),
    systemLanguage,
    regionalLanguage,
  };
}
